<a href="https://colab.research.google.com/github/j5675293/DESAFIOS__NPL_1/blob/NPL_1/DESAFIO_2_NLP_I.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# DESAFIO 2
## CARRERA DE ESPECIALIDAD EN INTELIGENCIA ARTIFICIAL - FIUBA
### MG. LIC. JUAN RUIZ OTONDO
#### a1702

El objetivo es utilizar documentos / corpus para crear embeddings de palabras basado en ese contexto. Se utilizará canciones de bandas para generar los embeddings, es decir, que los vectores tendrán la forma en función de como esa banda haya utilizado las palabras en sus canciones.

In [ ]:
!pip install gensim

  Using cached numpy-1.26.4-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (61 kB)
Using cached numpy-1.26.4-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (18.3 MB)
  Attempting uninstall: numpy
    Found existing installation: numpy 2.2.6
    Uninstalling numpy-2.2.6:
      Successfully uninstalled numpy-2.2.6
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
thinc 8.3.6 requires numpy<3.0.0,>=2.0.0, but you have numpy 1.26.4 which is incompatible.
tsfresh 0.21.0 requires scipy>=1.14.0; python_version >= "3.10", but you have scipy 1.13.1 which is incompatible.


In [ ]:
!pip install --upgrade numpy

  Using cached numpy-2.2.6-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (62 kB)
Using cached numpy-2.2.6-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (16.8 MB)
  Attempting uninstall: numpy
    Found existing installation: numpy 1.26.4
    Uninstalling numpy-1.26.4:
      Successfully uninstalled numpy-1.26.4
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
gensim 4.3.3 requires numpy<2.0,>=1.18.5, but you have numpy 2.2.6 which is incompatible.
tensorflow 2.18.0 requires numpy<2.1.0,>=1.26.0, but you have numpy 2.2.6 which is incompatible.
numba 0.60.0 requires numpy<2.1,>=1.22, but you have numpy 2.2.6 which is incompatible.
tsfresh 0.21.0 requires scipy>=1.14.0; python_version >= "3.10", but you have scipy 1.13.1 which is incompatible.


In [ ]:
!pip install --upgrade --force-reinstall gensim

  Using cached gensim-4.3.3-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (8.1 kB)
  Using cached numpy-1.26.4-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (61 kB)
  Using cached scipy-1.13.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (60 kB)
  Using cached smart_open-7.1.0-py3-none-any.whl.metadata (24 kB)
  Using cached wrapt-1.17.2-cp311-cp311-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (6.4 kB)
Using cached gensim-4.3.3-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (26.7 MB)
Using cached numpy-1.26.4-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (18.3 MB)
Using cached scipy-1.13.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (38.6 MB)
Using cached smart_open-7.1.0-py3-none-any.whl (61 kB)
Using cached wrapt-1.17.2-cp311-cp311-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (83 kB)
  Att

In [ ]:
!pip install numpy==1.21.6 gensim==4.1.2

ERROR: Ignored the following versions that require a different python version: 1.21.2 Requires-Python >=3.7,<3.11; 1.21.3 Requires-Python >=3.7,<3.11; 1.21.4 Requires-Python >=3.7,<3.11; 1.21.5 Requires-Python >=3.7,<3.11; 1.21.6 Requires-Python >=3.7,<3.11
ERROR: Could not find a version that satisfies the requirement numpy==1.21.6 (from versions: 1.3.0, 1.4.1, 1.5.0, 1.5.1, 1.6.0, 1.6.1, 1.6.2, 1.7.0, 1.7.1, 1.7.2, 1.8.0, 1.8.1, 1.8.2, 1.9.0, 1.9.1, 1.9.2, 1.9.3, 1.10.0.post2, 1.10.1, 1.10.2, 1.10.4, 1.11.0, 1.11.1, 1.11.2, 1.11.3, 1.12.0, 1.12.1, 1.13.0, 1.13.1, 1.13.3, 1.14.0, 1.14.1, 1.14.2, 1.14.3, 1.14.4, 1.14.5, 1.14.6, 1.15.0, 1.15.1, 1.15.2, 1.15.3, 1.15.4, 1.16.0, 1.16.1, 1.16.2, 1.16.3, 1.16.4, 1.16.5, 1.16.6, 1.17.0, 1.17.1, 1.17.2, 1.17.3, 1.17.4, 1.17.5, 1.18.0, 1.18.1, 1.18.2, 1.18.3, 1.18.4, 1.18.5, 1.19.0, 1.19.1, 1.19.2, 1.19.3, 1.19.4, 1.19.5, 1.20.0, 1.20.1, 1.20.2, 1.20.3, 1.21.0, 1.21.1, 1.22.0, 1.22.1, 1.22.2, 1.22.3, 1.22.4, 1.23.0, 1.23.1, 1.23.2, 1.23.3, 1.23

In [ ]:
# Se importa librerias iniciales para extraer dataset
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import multiprocessing
from gensim.models import Word2Vec

In [ ]:
# Se descarga la carpeta de dataset; proporcionado en clase
import os
import platform
if os.access('./songs_dataset', os.F_OK) is False:
    if os.access('songs_dataset.zip', os.F_OK) is False:
        if platform.system() == 'Windows':
            !curl https://raw.githubusercontent.com/FIUBA-Posgrado-Inteligencia-Artificial/procesamiento_lenguaje_natural/main/datasets/songs_dataset.zip -o songs_dataset.zip
        else:
            !wget songs_dataset.zip https://github.com/FIUBA-Posgrado-Inteligencia-Artificial/procesamiento_lenguaje_natural/raw/main/datasets/songs_dataset.zip
    !unzip -q songs_dataset.zip
else:
    print("El dataset ya se encuentra descargado")

--2025-05-20 01:43:18--  http://songs_dataset.zip/
Resolving songs_dataset.zip (songs_dataset.zip)... failed: Name or service not known.
wget: unable to resolve host address ‘songs_dataset.zip’
--2025-05-20 01:43:18--  https://github.com/FIUBA-Posgrado-Inteligencia-Artificial/procesamiento_lenguaje_natural/raw/main/datasets/songs_dataset.zip
Resolving github.com (github.com)... 140.82.114.4
Connecting to github.com (github.com)|140.82.114.4|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://raw.githubusercontent.com/FIUBA-Posgrado-Inteligencia-Artificial/procesamiento_lenguaje_natural/main/datasets/songs_dataset.zip [following]
--2025-05-20 01:43:18--  https://raw.githubusercontent.com/FIUBA-Posgrado-Inteligencia-Artificial/procesamiento_lenguaje_natural/main/datasets/songs_dataset.zip
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.109.133, 185.199.108.133, ...
Connecting to raw.githubusercontent.com (raw.

In [ ]:
# Posibles bandas a considerar
os.listdir("./songs_dataset/")

['johnny-cash.txt',
 'lady-gaga.txt',
 'eminem.txt',
 'r-kelly.txt',
 'Lil_Wayne.txt',
 'lil-wayne.txt',
 'kanye.txt',
 'janisjoplin.txt',
 'paul-simon.txt',
 'Kanye_West.txt',
 'michael-jackson.txt',
 'rihanna.txt',
 'dolly-parton.txt',
 'joni-mitchell.txt',
 'amy-winehouse.txt',
 'bjork.txt',
 'missy-elliott.txt',
 'alicia-keys.txt',
 'prince.txt',
 'jimi-hendrix.txt',
 'dickinson.txt',
 'beatles.txt',
 'dj-khaled.txt',
 'leonard-cohen.txt',
 'lorde.txt',
 'bob-marley.txt',
 'bob-dylan.txt',
 'dr-seuss.txt',
 'bruno-mars.txt',
 'notorious-big.txt',
 'al-green.txt',
 'nirvana.txt',
 'patti-smith.txt',
 'kanye-west.txt',
 'ludacris.txt',
 'notorious_big.txt',
 'drake.txt',
 'adele.txt',
 'lin-manuel-miranda.txt',
 'bieber.txt',
 'nursery_rhymes.txt',
 'nicki-minaj.txt',
 'bruce-springsteen.txt',
 'nickelback.txt',
 'disney.txt',
 'britney-spears.txt',
 'cake.txt',
 'radiohead.txt',
 'blink-182.txt']

In [ ]:
# Se procede al armado de dataset utilizando salto de línea para separar las oraciones/docs
df = pd.read_csv('songs_dataset/beatles.txt', sep='/n', header=None)
df.head()

<ipython-input-5-91c4bad52ba1>:2: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  df = pd.read_csv('songs_dataset/beatles.txt', sep='/n', header=None)


,0
0,"Yesterday, all my troubles seemed so far away"
1,Now it looks as though they're here to stay
2,"Oh, I believe in yesterday Suddenly, I'm not h..."
3,There's a shadow hanging over me.
4,"Oh, yesterday came suddenly Why she had to go ..."


In [ ]:
print("Cantidad de documentos:", df.shape[0])

Cantidad de documentos: 1846


# 1. Preprocesamiento

Se ejecuta lo siguiente:
*   Eliminación de caracteres especiales y números
*   Tokenización
*   Eliminación de stopwords en inglés
*   Lematización de palabras


In [ ]:
# Se importa las librerias necesarias
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import re
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from nltk.stem import WordNetLemmatizer
import gensim
from gensim.models import Word2Vec
from sklearn.manifold import TSNE
import plotly.express as px
import plotly.graph_objects as go
from tqdm.notebook import tqdm

In [ ]:
# Se descarga recursos de NLTK necesarios
print("Descargando recursos de NLTK...")
try:
    nltk.download('punkt')
    nltk.download('stopwords')
    nltk.download('wordnet')
    nltk.download('omw-1.4')  # Open Multilingual WordNet
except Exception as e:
    print(f"Error al descargar recursos NLTK: {e}")
    print("Intentando descarga alternativa...")
    import ssl
    try:
        _create_unverified_https_context = ssl._create_unverified_context
    except AttributeError:
        pass
    else:
        ssl._create_default_https_context = _create_unverified_https_context
    nltk.download('punkt')
    nltk.download('stopwords')
    nltk.download('wordnet')
    nltk.download('omw-1.4')  # Necesario para versiones recientes de NLTK


Descargando recursos de NLTK...


[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to /root/nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!


In [ ]:
# 1. Se carga y preprocesamiento de datos
print("Cargando datos...")
# Debe leer el archivo de canciones de los Beatles
try:
    # Lee el archivo como texto simple
    with open('songs_dataset/beatles.txt', 'r', encoding='utf-8') as file:
        lyrics = file.readlines()
except UnicodeDecodeError:
    # Intenta con otra codificación si falla la primera
    with open('songs_dataset/beatles.txt', 'r', encoding='latin1') as file:
        lyrics = file.readlines()

# Se crea DataFrame con las letras
df = pd.DataFrame(lyrics, columns=['lyrics'])

# Se limpia las líneas vacías y eliminar espacios en blanco al inicio/final
df['lyrics'] = df['lyrics'].str.strip()
df = df[df['lyrics'].str.len() > 0]

# Se muestra la información del dataset
print(f"Número total de canciones/documentos: {len(df)}")
print("\nPrimeras 5 canciones:")
print(df.head())


Cargando datos...
Número total de canciones/documentos: 1846

Primeras 5 canciones:
                                              lyrics
0      Yesterday, all my troubles seemed so far away
1        Now it looks as though they're here to stay
2  Oh, I believe in yesterday Suddenly, I'm not h...
3                  There's a shadow hanging over me.
4  Oh, yesterday came suddenly Why she had to go ...


In [ ]:
# 2. Preprocesamiento de dataset
def preprocess_text(text):
    """
    Función para preprocesar el texto:
    - Convertir a minúsculas
    - Eliminar caracteres especiales
    - Tokenizar (usando método simple para evitar dependencias)
    - Eliminar stopwords
    - Lematizar palabras
    """
    if isinstance(text, str):
        # Se convierte a minúsculas
        text = text.lower()

        # Se elimina caracteres especiales y números
        text = re.sub(r'[^\w\s]', ' ', text)
        text = re.sub(r'\d+', ' ', text)

        # Se tokeniza de manera simple (sin usar nltk.tokenize)
        tokens = text.split() # Reemplacé word_tokenize de NLTK por problemas de iteracción

        # Se elimina stopwords (aseguramos que stopwords esté disponible)
        try:
            stop_words = set(stopwords.words('english'))
        except LookupError:
            # Si falla, crea un conjunto básico de stopwords en inglés
            stop_words = {'a', 'an', 'the', 'and', 'or', 'but', 'if', 'because', 'as', 'what',
                         'when', 'where', 'how', 'who', 'which', 'this', 'that', 'these', 'those',
                         'then', 'just', 'so', 'than', 'such', 'both', 'through', 'about', 'for',
                         'is', 'are', 'was', 'were', 'be', 'been', 'being', 'have', 'has', 'had',
                         'having', 'do', 'does', 'did', 'doing', 'i', 'me', 'my', 'myself', 'we',
                         'our', 'ours', 'ourselves', 'you', "you're", "you've", "you'll", "you'd",
                         'your', 'yours', 'yourself', 'yourselves', 'he', 'him', 'his', 'himself',
                         'she', "she's", 'her', 'hers', 'herself', 'it', "it's", 'its', 'itself',
                         'they', 'them', 'their', 'theirs', 'themselves', 'to', 'from', 'up', 'down',
                         'in', 'out', 'on', 'off', 'over', 'under', 'again', 'further', 'then', 'once'}

        tokens = [word for word in tokens if word not in stop_words and len(word) > 2]

        # Se lematiza o simplemente usa las palabras como están si WordNetLemmatizer no está disponible
        try:
            lemmatizer = WordNetLemmatizer()
            tokens = [lemmatizer.lemmatize(word) for word in tokens]
        except LookupError:
            # Si no se puede lematizar, usar los tokens tal cual
            pass

        return tokens
    else:
        return []

print("\nPreprocesando textos...")
# Aplica preprocesamiento a todas las canciones
df['tokens'] = df['lyrics'].apply(preprocess_text)

# Elimina canciones que quedaron sin tokens después del preprocesamiento
df = df[df['tokens'].map(len) > 0]
print(f"Número de canciones después de eliminar documentos vacíos: {len(df)}")

# Ver cómo quedaron los tokens de las primeras canciones
print("\nEjemplo de tokenización (primeras 3 canciones):")
for i, tokens in enumerate(df['tokens'].head(3)):
    print(f"Canción {i+1}: {tokens[:15]}...")


Preprocesando textos...
Número de canciones después de eliminar documentos vacíos: 1834

Ejemplo de tokenización (primeras 3 canciones):
Canción 1: ['yesterday', 'trouble', 'seemed', 'far', 'away']...
Canción 2: ['look', 'though', 'stay']...
Canción 3: ['believe', 'yesterday', 'suddenly', 'half', 'man', 'used']...


# 2. Entrenamiento con Gensim
Se ejecuta lo siguiente:
*   Se utiliza Word2Vec de Gensim con arquitectura Skip-gram
*   Vectores de dimensión 100
*   Parámetros optimizados para corpus de canciones


In [ ]:
# 3. Entrenamiento de Word2Vec con Gensim
print("\nEntrenando modelo Word2Vec...")
# Parámetros del modelo
vector_size = 100  # Dimensionalidad de los vectores
window = 5         # Tamaño de la ventana contextual
min_count = 5      # Frecuencia mínima de palabras
workers = 4        # Número de hilos para entrenamiento paralelo
sg = 1             # 1 para skip-gram; 0 para CBOW

# Se entrena modelo
model = Word2Vec(
    df['tokens'].tolist(),
    vector_size=vector_size,
    window=window,
    min_count=min_count,
    workers=workers,
    sg=sg
)

# Información del modelo
print(f"Modelo entrenado con {len(model.wv.index_to_key)} palabras únicas")


Entrenando modelo Word2Vec...
Modelo entrenado con 315 palabras únicas


In [ ]:
# 4. Exploración de los embeddings
print("\nAnálisis de palabras similares:")

# Palabras relevantes para analizar (ejemplos de palabras frecuentes en canciones de los Beatles)
palabras_clave = ['love', 'girl', 'day', 'know', 'time', 'good', 'night']

# Encuentra las palabras más similares para cada palabra clave
for palabra in palabras_clave:
    try:
        similares = model.wv.most_similar(palabra, topn=5)
        print(f"\nPalabras más similares a '{palabra}':")
        for word, score in similares:
            print(f"{word}: {score:.4f}")
    except KeyError:
        print(f"La palabra '{palabra}' no está en el vocabulario del modelo")


Análisis de palabras similares:

Palabras más similares a 'love':
four: 0.9912
lonely: 0.9887
bang: 0.9882
three: 0.9878
money: 0.9877

Palabras más similares a 'girl':
love: 0.9872
money: 0.9870
bang: 0.9862
pepper: 0.9849
lonely: 0.9847

Palabras más similares a 'day':
lonely: 0.9837
little: 0.9834
say: 0.9834
dear: 0.9833
five: 0.9831

Palabras más similares a 'know':
four: 0.9885
little: 0.9881
like: 0.9879
say: 0.9879
money: 0.9878

Palabras más similares a 'time':
like: 0.9798
lonely: 0.9795
got: 0.9793
love: 0.9790
four: 0.9783

Palabras más similares a 'good':
little: 0.9799
say: 0.9794
bang: 0.9794
day: 0.9789
five: 0.9788

Palabras más similares a 'night':
little: 0.9878
dear: 0.9872
like: 0.9858
one: 0.9852
three: 0.9850


En el modelo a ejecutar se ha encontrando similitudes entre palabras con significados muy distintos. Por ejemplo, "love" y "four", "girl" y "money", "night" y "little". Esto sugiere que el corpus de entrenamiento puede estar influenciado por un contexto específico, como canciones, literatura, o algún conjunto de datos con asociaciones estilísticas particulares. Evaluaciones posteriores puede confirmar tal episodio.

# 3. Visualización de embeddings
Se ejecuta lo siguiente:
- Búsqueda de palabras similares para términos clave en canciones
- Visualización mediante t-SNE para reducir dimensionalidad
- Tres visualizaciones diferentes con Plotly:

 1. Mapa general de los embeddings de las palabras más frecuentes
 2. Análisis por categorías semánticas (emociones, tiempo, personas)
 3. Conexiones semánticas entre palabras clave




In [ ]:
# 5. Visualización de embeddings con t-SNE y Plotly
print("\nPreparando visualización t-SNE...")

# Se selecciona las palabras más frecuentes para visualizar (para evitar saturar el gráfico)
top_words = 150

# Se obtiene las palabras más frecuentes basadas en el modelo
words = model.wv.index_to_key[:top_words]

# Se extra los vectores para estas palabras
word_vectors = np.array([model.wv[word] for word in words])

# Se reduce dimensionalidad con t-SNE
tsne = TSNE(n_components=2, random_state=42, perplexity=30, n_iter=1000)
word_vectors_2d = tsne.fit_transform(word_vectors)

# Se crea un DataFrame para plotly
df_plot = pd.DataFrame({
    'word': words,
    'x': word_vectors_2d[:, 0],
    'y': word_vectors_2d[:, 1]
})

# Se visualiza con Plotly
fig = px.scatter(
    df_plot, x='x', y='y', text='word',
    title=f'Word Embeddings de letras de The Beatles - Top {top_words} palabras',
    labels={'x': 't-SNE 1', 'y': 't-SNE 2'}
)

# Se personaliza gráfico
fig.update_traces(
    textposition='top center',
    marker=dict(size=8, opacity=0.8),
    textfont=dict(size=10, color='darkblue')
)

fig.update_layout(
    height=700,
    width=900,
    hovermode='closest',
    template='plotly_white'
)

# Se muestra gráfico
fig.show()



Preparando visualización t-SNE...


/usr/local/lib/python3.11/dist-packages/sklearn/manifold/_t_sne.py:1164: FutureWarning: 'n_iter' was renamed to 'max_iter' in version 1.5 and will be removed in 1.7.
  warnings.warn(


- Los Word Embeddings no solo mapean palabras, sino que revelan la arquitectura conceptual subyacente en la música de The Beatles, se muestra cómo construyeron significado a través de la proximidad y relación entre conceptos.
- T-SNE 1 es un algoritmo de reducción de dimensionalidad que convierte datos de alta dimensión en un espacio visual.
- Este análisis demuestra que los embeddings capturan la "huella digital" lingüística de The Beatles, se muestra cómo:

  1. Creación de un universo semántico distintivo que los diferencia de otros artistas
  2. Integra temas universales (amor, tiempo, felicidad) de manera única
  3. Desarrolla patrones de co-ocurrencia que definen el estilo musical

# 4. Análisis de agrupaciones de palabras

In [ ]:
# 6. Análisis de agrupaciones de palabras
print("\nAnalizando agrupaciones por categorías...")

# Se define categorías de palabras para analizar
categories = {
    'Emociones': ['love', 'feel', 'heart', 'cry', 'happy', 'sad'],
    'Tiempo': ['day', 'night', 'time', 'year', 'today', 'tomorrow'],
    'Personas': ['girl', 'boy', 'man', 'woman', 'friend', 'baby'],
}

# se crea visualización de clusters por categoría
fig_categories = go.Figure()

# Colores para cada categoría
colors = ['#FF5733', '#33BBFF', '#33FF57']

# Se agrega cada categoría al gráfico
for i, (category, cat_words) in enumerate(categories.items()):
    # Se filtra palabras que están en el vocabulario
    valid_words = [word for word in cat_words if word in model.wv]

    if valid_words:
        # Se obtiene vectores para estas palabras
        cat_vectors = np.array([model.wv[word] for word in valid_words])

        # Se aplica t-SNE a estos vectores
        cat_tsne = TSNE(n_components=2, random_state=42, perplexity=min(5, len(valid_words)-1) if len(valid_words) > 1 else 1)
        cat_vectors_2d = cat_tsne.fit_transform(cat_vectors)

        # Se agrega al gráfico
        for j, word in enumerate(valid_words):
            fig_categories.add_trace(
                go.Scatter(
                    x=[cat_vectors_2d[j, 0]],
                    y=[cat_vectors_2d[j, 1]],
                    mode='markers+text',
                    name=word,
                    text=word,
                    textposition='top center',
                    marker=dict(size=10, color=colors[i % len(colors)]),
                    legendgroup=category,
                    hoverinfo='text',
                    hovertext=f"{word} ({category})"
                )
            )

# Layout para el gráfico de categorías
fig_categories.update_layout(
    title='Agrupación de palabras por categorías semánticas en canciones de The Beatles',
    height=600,
    width=800,
    template='plotly_white',
    legend_title_text='Palabras',
    hovermode='closest'
)

# Se muestra gráfico de categorías
fig_categories.show()



Analizando agrupaciones por categorías...


Este análisis sugiere que en las letras de The Beatles, las palabras tienden a agruparse en torno a los siguientes temas semánticos principales:

- Amor y emociones (Positivas y Negativas),  un núcleo fuerte de palabras emocionales, con el amor y la felicidad como el centro más denso, y la tristeza/llanto como un subgrupo.
- Tiempo y temporalidad, una clara concentración de palabras relacionadas con el tiempo, indicando una posible preocupación por el paso del tiempo, momentos específicos, o el día a día.
- Relaciones personales, un grupo de palabras que denotan roles o relaciones entre personas, aunque estas parecen tener una conexión más difusa con los otros temas, lo que podría significar que estas relaciones se entrelazan con una variedad de emociones y temporalidades.

# 5. Análisis de conexiones semanticas entre palabras

In [ ]:
# 7. Análisis de conexiones semánticas entre palabras
print("\nAnalizando conexiones semánticas...")

# Palabras principales para analizar
central_words = ['love', 'girl', 'day']

# Se crea gráfico de conexiones
fig_connections = go.Figure()

# Para cada palabra central, mostrar sus palabras más similares
for central_word in central_words:
    if central_word in model.wv:
        # Se obtiene las palabras más similares
        similar_words = [word for word, _ in model.wv.most_similar(central_word, topn=7)]

        # Se obtiene vectores 2D para estas palabras
        words_to_plot = [central_word] + similar_words
        vectors = np.array([model.wv[word] for word in words_to_plot])

        # Se aplica t-SNE
        tsne_conn = TSNE(n_components=2, random_state=42, perplexity=min(5, len(words_to_plot)-1))
        vectors_2d = tsne_conn.fit_transform(vectors)

        # Se agrega nodo central
        fig_connections.add_trace(
            go.Scatter(
                x=[vectors_2d[0, 0]],
                y=[vectors_2d[0, 1]],
                mode='markers+text',
                marker=dict(size=15, color='red'),
                text=central_word,
                name=central_word,
                textposition='middle center',
                textfont=dict(size=12, color='white')
            )
        )

        # Se agrega nodos conectados
        fig_connections.add_trace(
            go.Scatter(
                x=vectors_2d[1:, 0],
                y=vectors_2d[1:, 1],
                mode='markers+text',
                marker=dict(size=10),
                text=similar_words,
                name=f'Relacionadas con {central_word}',
                textposition='top center'
            )
        )

        # Se agrega líneas de conexión
        for i in range(1, len(vectors_2d)):
            fig_connections.add_shape(
                type="line",
                x0=vectors_2d[0, 0],
                y0=vectors_2d[0, 1],
                x1=vectors_2d[i, 0],
                y1=vectors_2d[i, 1],
                line=dict(color="lightgrey", width=1)
            )

# Layout para el gráfico de conexiones
fig_connections.update_layout(
    title='Conexiones semánticas entre palabras clave en canciones de The Beatles',
    height=600,
    width=800,
    template='plotly_white',
    showlegend=True
)

# Se muestra gráfico de conexiones
fig_connections.show()


Analizando conexiones semánticas...


- Temas centrales,  "Love", "girl" y "day" son conceptos fundamentales en las canciones de The Beatles.
- Algunas palabras, como "money" y "lonely", tiende a ser conectores importantes, apareciendo en el ámbito semántico de múltiples palabras clave centrales.
- Los puntos con más líneas saliendo de ellos son más "centrales" en la red de conexiones.

In [ ]:
# Guarda el modelo para uso futuro
print("\nGuardando modelo...")
model.save("beatles_word2vec.model")
print("Modelo guardado como 'beatles_word2vec.model'")
print("\n¡Análisis completo! Explora los gráficos para ver las relaciones entre palabras en las canciones de The Beatles.")


Guardando modelo...
Modelo guardado como 'beatles_word2vec.model'

¡Análisis completo! Explora los gráficos para ver las relaciones entre palabras en las canciones de The Beatles.


# ANALISIS COMPARATIVO
# Entrenamiento con spaCy

#### Se utiliza spaCy, que ofrece un procesamiento lingüístico más avanzado y preciso que NLTK. Este código permite comparar los resultados entre ambos enfoques. El desarrollo de código tiene base estructural a NLTK.

In [ ]:
# Se importa las librerias requeridos
import spacy
from tqdm.notebook import tqdm
import gensim
from gensim.models import Word2Vec
from sklearn.manifold import TSNE
import plotly.express as px
import plotly.graph_objects as go

In [ ]:
# Se instala spaCy
!pip install -U spacy
!python -m spacy download en_core_web_sm

  Using cached numpy-2.2.6-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (62 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 30.6/30.6 MB 43.4 MB/s eta 0:00:00
Using cached numpy-2.2.6-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (16.8 MB)
  Attempting uninstall: numpy
    Found existing installation: numpy 1.26.4
    Uninstalling numpy-1.26.4:
      Successfully uninstalled numpy-1.26.4
  Attempting uninstall: spacy
    Found existing installation: spacy 3.8.5
    Uninstalling spacy-3.8.5:
      Successfully uninstalled spacy-3.8.5
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
gensim 4.3.3 requires numpy<2.0,>=1.18.5, but you have numpy 2.2.6 which is incompatible.
tensorflow 2.18.0 requires numpy<2.1.0,>=1.26.0, but you have numpy 2.2.6 which is incompatible.
numba 0.60.0 requires numpy<2.1,>=1.22, but you have numpy 2.2.

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.8/12.8 MB 95.5 MB/s eta 0:00:00
✔ Download and installation successful
You can now load the package via spacy.load('en_core_web_sm')
⚠ Restart to reload dependencies
If you are in a Jupyter or Colab notebook, you may need to restart Python in
order to load all the package's dependencies. You can do this by selecting the
'Restart kernel' or 'Restart runtime' option.


In [ ]:
# 1. Carga de datos
print("Cargando datos...")
try:
    # Leer el archivo como texto simple
    with open('songs_dataset/beatles.txt', 'r', encoding='utf-8') as file:
        lyrics = file.readlines()
except UnicodeDecodeError:
    # Intentar con otra codificación si falla la primera
    with open('songs_dataset/beatles.txt', 'r', encoding='latin1') as file:
        lyrics = file.readlines()

# Crear DataFrame con las letras
df = pd.DataFrame(lyrics, columns=['lyrics'])

# Limpiar las líneas vacías y eliminar espacios en blanco al inicio/final
df['lyrics'] = df['lyrics'].str.strip()
df = df[df['lyrics'].str.len() > 0]

# Mostrar la información del dataset
print(f"Número total de canciones/documentos: {len(df)}")
print("\nPrimeras 5 canciones:")
print(df.head())

Cargando datos...
Número total de canciones/documentos: 1846

Primeras 5 canciones:
                                              lyrics
0      Yesterday, all my troubles seemed so far away
1        Now it looks as though they're here to stay
2  Oh, I believe in yesterday Suddenly, I'm not h...
3                  There's a shadow hanging over me.
4  Oh, yesterday came suddenly Why she had to go ...


# 1. Preprocesamiento

In [ ]:
# 2. Preprocesamiento con spaCy
print("\nCargando modelo de spaCy...")
nlp = spacy.load("en_core_web_sm")

def preprocess_with_spacy(text):
    """
    Función para preprocesar el texto usando spaCy:
    - Parseo con spaCy
    - Eliminación de stopwords, puntuación y números
    - Lematización
    """
    if isinstance(text, str):
        # se procesa el texto con spaCy
        doc = nlp(text.lower())

        # Se filtra tokens: eliminar stopwords, puntuación, espacios y números
        # También se eliminan palabras muy cortas (longitud < 3)
        tokens = [token.lemma_ for token in doc
                if not token.is_stop
                and not token.is_punct
                and not token.is_space
                and not token.like_num
                and len(token.text) > 2]

        return tokens
    else:
        return []

# Se procesa las letras en lotes pequeños para evitar problemas de memoria
print("\nPreprocesando textos con spaCy...")
batch_size = 100
num_batches = (len(df) + batch_size - 1) // batch_size

tokens_list = []
for i in tqdm(range(num_batches)):
    start_idx = i * batch_size
    end_idx = min((i + 1) * batch_size, len(df))
    batch = df['lyrics'][start_idx:end_idx]

    batch_tokens = [preprocess_with_spacy(text) for text in batch]
    tokens_list.extend(batch_tokens)

df['tokens'] = tokens_list

# Se elimina canciones que quedaron sin tokens después del preprocesamiento
df = df[df['tokens'].map(len) > 0]
print(f"Número de canciones después de eliminar documentos vacíos: {len(df)}")

# Se observa cómo quedaron los tokens de las primeras canciones
print("\nEjemplo de tokenización con spaCy (primeras 3 canciones):")
for i, tokens in enumerate(df['tokens'].head(3)):
    print(f"Canción {i+1}: {tokens[:15]}...")


Cargando modelo de spaCy...

Preprocesando textos con spaCy...


  0%|          | 0/19 [00:00<?, ?it/s]

Número de canciones después de eliminar documentos vacíos: 1761

Ejemplo de tokenización con spaCy (primeras 3 canciones):
Canción 1: ['yesterday', 'trouble', 'far', 'away']...
Canción 2: ['look', 'stay']...
Canción 3: ['believe', 'yesterday', 'suddenly', 'half', 'man']...


In [ ]:
# 3. Entrenamiento de Word2Vec
print("\nEntrenando modelo Word2Vec...")
# Parámetros del modelo
vector_size = 100  # Dimensionalidad de los vectores
window = 5         # Tamaño de la ventana contextual
min_count = 5      # Frecuencia mínima de palabras
workers = 4        # Número de hilos para entrenamiento paralelo
sg = 1             # 1 para skip-gram; 0 para CBOW

# Entrenamiento de modelo
model = Word2Vec(
    df['tokens'].tolist(),
    vector_size=vector_size,
    window=window,
    min_count=min_count,
    workers=workers,
    sg=sg
)

# Información del modelo a entrenado
print(f"Modelo entrenado con {len(model.wv.index_to_key)} palabras únicas")


Entrenando modelo Word2Vec...
Modelo entrenado con 259 palabras únicas


In [ ]:
# 4. Exploración embeddings
print("\nAnálisis de palabras similares:")

# Palabras relevantes para analizar
palabras_clave = ['love', 'girl', 'day', 'know', 'time', 'good', 'night']

# Encuentra las palabras más similares para cada palabra clave
for palabra in palabras_clave:
    try:
        similares = model.wv.most_similar(palabra, topn=5)
        print(f"\nPalabras más similares a '{palabra}':")
        for word, score in similares:
            print(f"{word}: {score:.4f}")
    except KeyError:
        print(f"La palabra '{palabra}' no está en el vocabulario del modelo")


Análisis de palabras similares:

Palabras más similares a 'love':
prudence: 0.9459
bang: 0.9384
day: 0.9362
eye: 0.9360
buy: 0.9356

Palabras más similares a 'girl':
band: 0.9228
heart: 0.9216
little: 0.9207
prudence: 0.9196
go: 0.9194

Palabras más similares a 'day':
prudence: 0.9409
helloa: 0.9380
love: 0.9362
pepper: 0.9356
maxwell: 0.9344

Palabras más similares a 'know':
bang: 0.9407
go: 0.9333
night: 0.9330
lonely: 0.9317
like: 0.9307

Palabras más similares a 'time':
little: 0.8808
band: 0.8788
prudence: 0.8750
hey: 0.8722
lonely: 0.8721

Palabras más similares a 'good':
prudence: 0.9176
buy: 0.9164
love: 0.9098
bang: 0.9037
go: 0.9023

Palabras más similares a 'night':
eye: 0.9445
like: 0.9433
little: 0.9414
prudence: 0.9403
pepper: 0.9403


- Este análisis de "palabras más similares" se basa en las incrustaciones de palabras (Word Embeddings), que representan el significado de las palabras en un espacio vectorial. La puntuación (cercana a 1.0) indica la similitud de coseno, es decir, qué tan cerca están los vectores de esas palabras en el espacio semántico.
- El análisis reafirma la importancia de temas como el amor, las relaciones humanas ("girl", "boy"), el tiempo ("day", "night", "time"), y las emociones ("lonely", "heart") en sus letras que inicialmente se habia descrito.

In [ ]:
# 5. Visualización de embeddings con t-SNE y Plotly
print("\nPreparando visualización t-SNE...")

# Se selecciona las palabras más frecuentes para visualizar
top_words = 150

# Se obtiene las palabras más frecuentes basadas en el modelo
words = model.wv.index_to_key[:top_words]

# Se extrae los vectores para estas palabras
word_vectors = np.array([model.wv[word] for word in words])

# Se reduce dimensionalidad con t-SNE
tsne = TSNE(n_components=2, random_state=42, perplexity=30, n_iter=1000)
word_vectors_2d = tsne.fit_transform(word_vectors)

# Se crea un DataFrame para plotly
df_plot = pd.DataFrame({
    'word': words,
    'x': word_vectors_2d[:, 0],
    'y': word_vectors_2d[:, 1]
})

# Se visualiza con Plotly
fig = px.scatter(
    df_plot, x='x', y='y', text='word',
    title=f'Word Embeddings (spaCy) de letras de The Beatles - Top {top_words} palabras',
    labels={'x': 't-SNE 1', 'y': 't-SNE 2'}
)

# Personalizacion de  gráfico
fig.update_traces(
    textposition='top center',
    marker=dict(size=8, opacity=0.8),
    textfont=dict(size=10, color='darkblue')
)

fig.update_layout(
    height=700,
    width=900,
    hovermode='closest',
    template='plotly_white'
)

# Se muestra grafico plotly
fig.show()


Preparando visualización t-SNE...


/usr/local/lib/python3.11/dist-packages/sklearn/manifold/_t_sne.py:1164: FutureWarning: 'n_iter' was renamed to 'max_iter' in version 1.5 and will be removed in 1.7.
  warnings.warn(


- El gráfico muestra un mapa semántico donde las palabras se agrupan en función de sus contextos de aparición y significado en las letras de The Beatles. Cada punto es una palabra.
- La disposición de las palabras valida la eficacia de los Word Embeddings y t-SNE para capturar relaciones semánticas. Las palabras con significados similares (ej. "love", "feel", "happy" o "sad", "lonely") están consistentemente agrupadas.

In [ ]:
# 6. Análisis de agrupaciones de palabras
print("\nAnalizando agrupaciones por categorías...")

# Se define categorías de palabras para analizar
categories = {
    'Emociones': ['love', 'feel', 'heart', 'cry', 'happy', 'sad'],
    'Tiempo': ['day', 'night', 'time', 'year', 'today', 'tomorrow'],
    'Personas': ['girl', 'boy', 'man', 'woman', 'friend', 'baby'],
}

# Se crea visualización de clusters por categoría
fig_categories = go.Figure()

# Colores para cada categoría
colors = ['#FF5733', '#33BBFF', '#33FF57']

# Se agrega cada categoría al gráfico
for i, (category, cat_words) in enumerate(categories.items()):
    # Se filtra palabras que están en el vocabulario
    valid_words = [word for word in cat_words if word in model.wv]

    if valid_words:
        # Se obtiene vectores para estas palabras
        cat_vectors = np.array([model.wv[word] for word in valid_words])

        # Se aplica t-SNE a estos vectores
        cat_tsne = TSNE(n_components=2, random_state=42, perplexity=min(5, len(valid_words)-1) if len(valid_words) > 1 else 1)
        cat_vectors_2d = cat_tsne.fit_transform(cat_vectors)

        # Se agrega al gráfico
        for j, word in enumerate(valid_words):
            fig_categories.add_trace(
                go.Scatter(
                    x=[cat_vectors_2d[j, 0]],
                    y=[cat_vectors_2d[j, 1]],
                    mode='markers+text',
                    name=word,
                    text=word,
                    textposition='top center',
                    marker=dict(size=10, color=colors[i % len(colors)]),
                    legendgroup=category,
                    hoverinfo='text',
                    hovertext=f"{word} ({category})"
                )
            )

# Layout para el gráfico de categorías
fig_categories.update_layout(
    title='Agrupación de palabras por categorías semánticas (spaCy) en canciones de The Beatles',
    height=600,
    width=800,
    template='plotly_white',
    legend_title_text='Palabras',
    hovermode='closest'
)

# Se muestra gráfico de categorías
fig_categories.show()



Analizando agrupaciones por categorías...


- Los puntos representan palabras, y la proximidad entre ellos indica similitud semántica. Los colores asignados a los puntos representan categorías semánticas (emociones, tiempo, personas). Sin embargo se evidencia que estan disperson, esto puede ser por numero de corpus.


In [ ]:
# 7. Análisis de conexiones semánticas entre palabras
print("\nAnalizando conexiones semánticas...")

# Palabras principales para analizar
central_words = ['love', 'girl', 'day']

# Se crea gráfico de conexiones
fig_connections = go.Figure()

# Para cada palabra central, se muestra sus palabras más similares
for central_word in central_words:
    if central_word in model.wv:
        # Obtiene las palabras más similares
        similar_words = [word for word, _ in model.wv.most_similar(central_word, topn=7)]

        # Obtiene vectores 2D para estas palabras
        words_to_plot = [central_word] + similar_words
        vectors = np.array([model.wv[word] for word in words_to_plot])

        # Aplicar t-SNE
        tsne_conn = TSNE(n_components=2, random_state=42, perplexity=min(5, len(words_to_plot)-1))
        vectors_2d = tsne_conn.fit_transform(vectors)

        # Agrega nodo central
        fig_connections.add_trace(
            go.Scatter(
                x=[vectors_2d[0, 0]],
                y=[vectors_2d[0, 1]],
                mode='markers+text',
                marker=dict(size=15, color='red'),
                text=central_word,
                name=central_word,
                textposition='middle center',
                textfont=dict(size=12, color='white')
            )
        )

        # Agrega nodos conectados
        fig_connections.add_trace(
            go.Scatter(
                x=vectors_2d[1:, 0],
                y=vectors_2d[1:, 1],
                mode='markers+text',
                marker=dict(size=10),
                text=similar_words,
                name=f'Relacionadas con {central_word}',
                textposition='top center'
            )
        )

        # Agrega líneas de conexión
        for i in range(1, len(vectors_2d)):
            fig_connections.add_shape(
                type="line",
                x0=vectors_2d[0, 0],
                y0=vectors_2d[0, 1],
                x1=vectors_2d[i, 0],
                y1=vectors_2d[i, 1],
                line=dict(color="lightgrey", width=1)
            )

# Layout para el gráfico de conexiones
fig_connections.update_layout(
    title='Conexiones semánticas entre palabras clave (spaCy) en canciones de The Beatles',
    height=600,
    width=800,
    template='plotly_white',
    showlegend=True
)

# el grafico
fig_connections.show()


Analizando conexiones semánticas...


- Se evidencia múltiples instancias de "love" (rojo), "girl" (violeta) y "day" (cian) dispersas en el gráfico.
- Se confirma que el amor, las relaciones personales (girl, boy) y el tiempo son los pilares semánticos de sus letras.


# Comparación NLTK y spaCy

In [ ]:
# 8. Comparación con NLTK
try:
    # Carga el modelo NLTK
    nltk_model = Word2Vec.load("beatles_word2vec.model")

    print("\nComparando resultados entre spaCy y NLTK:")

    # Guarda el modelo spaCy para referencia
    model.save("beatles_spacy_word2vec.model")
    print("Modelo spaCy guardado como 'beatles_spacy_word2vec.model'")

    # Compara palabras similares para términos clave
    for palabra in ['love', 'girl']:
        if palabra in model.wv and palabra in nltk_model.wv:
            print(f"\nPalabras similares a '{palabra}':")

            print("Según spaCy:")
            for word, score in model.wv.most_similar(palabra, topn=5):
                print(f"  {word}: {score:.4f}")

            print("\nSegún NLTK:")
            for word, score in nltk_model.wv.most_similar(palabra, topn=5):
                print(f"  {word}: {score:.4f}")

    # Visualización comparativa
    # Crea un DataFrame para visualizar ambos modelos
    common_words = [w for w in model.wv.index_to_key[:100] if w in nltk_model.wv][:50]

    if common_words:
        # Crea un gráfico comparativo
        fig_compare = go.Figure()

        # Agrega puntos para el modelo spaCy
        spacy_vectors = np.array([model.wv[word] for word in common_words])
        spacy_tsne = TSNE(n_components=2, random_state=42, perplexity=min(30, len(common_words)-1))
        spacy_2d = spacy_tsne.fit_transform(spacy_vectors)

        fig_compare.add_trace(
            go.Scatter(
                x=spacy_2d[:, 0],
                y=spacy_2d[:, 1],
                mode='markers+text',
                marker=dict(size=8, color='blue'),
                text=common_words,
                name='spaCy',
                textposition='top center'
            )
        )

        # Agrega puntos para el modelo NLTK
        nltk_vectors = np.array([nltk_model.wv[word] for word in common_words])
        nltk_tsne = TSNE(n_components=2, random_state=42, perplexity=min(30, len(common_words)-1))
        nltk_2d = nltk_tsne.fit_transform(nltk_vectors)

        fig_compare.add_trace(
            go.Scatter(
                x=nltk_2d[:, 0],
                y=nltk_2d[:, 1],
                mode='markers+text',
                marker=dict(size=8, color='red'),
                text=common_words,
                name='NLTK',
                textposition='top center'
            )
        )

        fig_compare.update_layout(
            title='Comparación de embeddings entre spaCy y NLTK',
            height=800,
            width=1200,
            template='plotly_white'
        )

        fig_compare.show()

except Exception as e:
    print(f"No se pudo realizar la comparación con NLTK: {e}")
    # Se guarda el modelo spaCy
    model.save("beatles_spacy_word2vec.model")
    print("Modelo spaCy guardado como 'beatles_spacy_word2vec.model'")

print("\n¡Análisis con spaCy completo! Explora los gráficos para ver las relaciones entre palabras en las canciones de The Beatles.")


Comparando resultados entre spaCy y NLTK:
Modelo spaCy guardado como 'beatles_spacy_word2vec.model'

Palabras similares a 'love':
Según spaCy:
  prudence: 0.9459
  bang: 0.9384
  day: 0.9362
  eye: 0.9360
  buy: 0.9356

Según NLTK:
  four: 0.9912
  lonely: 0.9887
  bang: 0.9882
  three: 0.9878
  money: 0.9877

Palabras similares a 'girl':
Según spaCy:
  band: 0.9228
  heart: 0.9216
  little: 0.9207
  prudence: 0.9196
  go: 0.9194

Según NLTK:
  love: 0.9872
  money: 0.9870
  bang: 0.9862
  pepper: 0.9849
  lonely: 0.9847



¡Análisis con spaCy completo! Explora los gráficos para ver las relaciones entre palabras en las canciones de The Beatles.


## Conclusión comparativo
- NLTK se inclina más hacia las co-ocurrencias numéricas y temáticas directas ("money", "lonely"), mientras que spaCy es más amplio en sus asociaciones contextuales, incluyendo referencias a canciones ("Prudence") y conceptos abstractos/sensoriales ("day", "eye"). Las puntuaciones de NLTK son altas que resultan por una posible sobreajuste o un corpus demasiado pequeño/parámetros no óptimos que hacen que las palabras se "peguen" demasiado entre sí.
- NLTK conecta "girl" directamente con los temas centrales de amor, dinero y soledad, y con nombres propios como "pepper". spaCy, por otro lado, ofrece asociaciones más "descriptivas" o "modales" ("little", "heart", "band", "go"), además de la referencia a "Prudence".
- Las puntuaciones de spaCy, aunque también altas, son más "realistas" para la similitud de coseno en Word Embeddings (donde 0.94 ya es muy alto).
- NLTK parece enfocarse más en las co-ocurrencias directas y temáticas fuertes, entre tanto spaCy parece capturar una gama más matizada y contextual de similitude.
- Las similitudes más bajas pero más diversas de spaCy podrían indicar un modelo de embeddings más robusto y generalizable dentro del corpus de los Beatles, que captura una variedad de contextos y no solo las co-ocurrencias más obvias o frecuentes.
- Las similitudes extremadamente altas de NLTK podrían indicar un modelo más sensible a las frecuencias exactas o patrones específicos en el corpus, lo que podría llevar a sobreajuste.
- Se podria sugerir simular con paramétros de entrenamiento mas elevados

- La característica más impactante es la clara separación espacial entre los embeddings generados por spaCy (azules) y NLTK (rojos). Esta separación indica que los dos modelos han aprendido representaciones vectoriales fundamentalmente diferentes para las mismas palabras, incluso si el corpus y el algoritmo (Word2Vec) son los mismos. Sus espacios de significado son distintos.